src/calculus/moreauEnvelope.jl

```diff
@@ -33,17 +28,15 @@ is_quadratic(f::MoreauEnvelope) = is_generalized_quadratic(f.g)
 is_strongly_convex(f::MoreauEnvelope) = is_strongly_convex(f.g)
 
 function (f::MoreauEnvelope)(x::AbstractArray)
-	if isnull(f.buf[1])
-		f.buf[1] = Nullable{AbstractArray}(similar(x))
-	end
-	g_prox = prox!(get(f.buf[1]), f.g, x, f.lambda)
-	return g_prox + 1/(2*f.lambda)*deepvecnorm(get(f.buf[1])-x)^2
+	buf = similar(x)
+	g_prox = prox!(buf, f.g, x, f.lambda)
+	return g_prox + 1/(2*f.lambda)*norm(buf .- x)^2
 end
 
 function gradient!(grad::AbstractArray, f::MoreauEnvelope, x::AbstractArray)
 	g_prox = prox!(grad, f.g, x, f.lambda)
-	grad .= (x - grad)/f.lambda
-	fx = g_prox + (f.lambda/2)*deepvecnorm(grad)^2
+	grad .= (x .- grad)/f.lambda
```

Comment:
I think that `./` might be more efficient here?
Example:
```julia
f1(x,y,z) = x .= (x .- y)/z
f2(x,y,z) = x .= (x .- y)/.z
x = randn(1000); y = randn(1000); z = 2.0
```
```julia
julia> @code_lowered f1(x,y,z)
CodeInfo(
1 1 ─ %1 = Base.Broadcast.materialize!                                                                            │
  │   %2 = Base.Broadcast.broadcasted                                                                             │
  │   %3 = Base.Broadcast.materialize                                                                             │
  │   %4 = Base.Broadcast.broadcasted                                                                             │
  │   %5 = (%4)(Main.:-, x, y)                                                                                    │
  │   %6 = (%3)(%5)                                                                                               │
  │   %7 = %6 / z                                                                                                 │
  │   %8 = (%2)(Base.identity, %7)                                                                                │
  │   %9 = (%1)(x, %8)                                                                                            │
  └──      return %9                                                                                              │
)
```
```julia
julia> @code_lowered f2(x,y,z)
CodeInfo(
1 1 ─ %1 = Base.Broadcast.materialize!                                                                            │
  │   %2 = Base.Broadcast.broadcasted                                                                             │
  │   %3 = Base.Broadcast.broadcasted                                                                             │
  │   %4 = (%3)(Main.:-, x, y)                                                                                    │
  │   %5 = (%2)(Main.:/, %4, z)                                                                                   │
  │   %6 = (%1)(x, %5)                                                                                            │
  └──      return %6                                                                                              │
)

```

where we see that `f1` has tw materialize twice (tested on julia 0.7, with 2x difference in time).

Reply:
Good catch! Fixing this